packages/flutter/lib/src/material/scaffold.dart

```diff
@@ -1440,89 +1478,206 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin {
+  Color _bodyScrimColor;
 
-  void _maybeBuildCurrentBottomSheet() {
-    if (widget.bottomSheet != null) {
+  void _maybeBuildPersistentBottomSheet() {
+    if (widget.bottomSheet != null && _currentBottomSheet == null) {
       // The new _currentBottomSheet is not a local history entry so a "back" button
       // will not be added to the Scaffold's appbar and the bottom sheet will not
       // support drag or swipe to dismiss.
+      AnimationController animationController;
+      if (!widget.bottomSheetIsScrollControlled) {
+        animationController = BottomSheet.createAnimationController(this)..value = 1.0;
+      }
       _currentBottomSheet = _buildBottomSheet<void>(
         (BuildContext context) => widget.bottomSheet,
-        BottomSheet.createAnimationController(this) ..value = 1.0,
-        false,
+        true,
+        animationController: animationController,
+        isScrollControlled: widget.bottomSheetIsScrollControlled,
       );
     }
   }
 
   void _closeCurrentBottomSheet() {
     if (_currentBottomSheet != null) {
       _currentBottomSheet.close();
-      assert(_currentBottomSheet == null);
+      assert(() {
+        _currentBottomSheet?._completer?.future?.whenComplete(() {
+          assert(_currentBottomSheet == null);
+        });
+        return true;
+      }());
     }
   }
 
-  PersistentBottomSheetController<T> _buildBottomSheet<T>(WidgetBuilder builder, AnimationController controller, bool isLocalHistoryEntry) {
+  PersistentBottomSheetController<T> _buildBottomSheet<T>(
+      WidgetBuilder builder,
+      bool isPersistent, {
+      double initialHeight = 0.5,
+      AnimationController animationController,
+      bool isScrollControlled = false,
+    }) {
+    assert(() {
+      if (widget.bottomSheet != null && isPersistent && _currentBottomSheet != null) {
+        throw FlutterError(
+          'Scaffold.bottomSheet cannot be specified while a bottom sheet displayed '
+          'with showBottomSheet() is still visible.\n Rebuild the Scaffold with a null '
+          'bottomSheet before calling showBottomSheet().'
+        );
+      }
+      return true;
+    }());
+
+    LocalHistoryEntry persistentSheetHistoryEntry;
+    if (!isScrollControlled) {
+      _bottomSheetScrollController = null;
+    } else {
+      _bottomSheetScrollController = BottomSheet.createScrollController(
+        initialHeightPercentage: initialHeight,
+        minTop: isPersistent ? initialHeight : 0.0,
+        isPersistent: isPersistent,
+        context: context,
+      )..addTopListener(() => setState(() {
+        // If we're very close to the top, just set the scale to 0.
+        // This avoids issues where a fling can send us up so fast we don't acutally
+        // fully dismiss the fab.
+        if (_bottomSheetScrollController.top < 2) {
+          floatingActionButtonVisibilityValue = 0.0;
+        } else {
+          final double screenHeight = MediaQuery.of(context).size.height;
```

Comment:
Shouldn't depend on screenHeight...

Reply:
I'm not sure how to resolve this - maybe we can talk more tomorrow.  The problem is finding the correct height to use during build.